stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,246 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SOLVER_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SOLVER_HPP
+
+#include <stan/math/rev/mat/functor/algebra_system.hpp>
+#include <stan/math/prim/mat/fun/mdivide_left.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/scal/meta/is_var.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/err/check_consistent_size.hpp>
+#include <stan/math/prim/arr/err/check_nonzero_size.hpp>
+#include <unsupported/Eigen/NonLinearOptimization>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * The vari class for the algebraic solver. We compute the  Jacobian of
+     * the solutions with respect to the parameters using the implicit
+     * function theorem. The call to Jacobian() occurs outside the call to
+     * chain() -- this prevents malloc issues.
+     *
+     * Members:
+     * y_ the vector of parameters
+     * y_size_ the number of parameters
+     * x_size_ the number of unknowns
+     * theta_ the vector of solution
+     * Jx_y_ the Jacobian of the solution with respect to the parameters.
+     */
+    template <typename FS, typename F, typename T, typename FX>
+    struct algebra_solver_vari : public vari {
+      vari** y_;
+      int y_size_;
+      int x_size_;
+      vari** theta_;
+      Eigen::MatrixXd Jx_y_;
+
+      algebra_solver_vari(const FS& fs,
+                          const F& f,
+                          const Eigen::VectorXd x,
+                          const Eigen::Matrix<T, Eigen::Dynamic, 1> y,
+                          const std::vector<double> dat,
+                          const std::vector<int> dat_int,
+                          const Eigen::VectorXd theta_dbl,
+                          FX& fx,
+                          std::ostream* msgs)
+        : vari(theta_dbl(0)),
+          y_(ChainableStack::memalloc_.alloc_array<vari*>(y.size())),
+          y_size_(y.size()),
+          x_size_(x.size()),
+          theta_(ChainableStack::memalloc_.alloc_array<vari*>(x.size())) {
+        for (int i = 0; i < y.size(); ++i)
+          y_[i] = y(i).vi_;
+
+        theta_[0] = this;
+        for (int i = 1; i < x.size(); ++i)
+          theta_[i] = new vari(theta_dbl(i), false);
+
+        // Compute the Jacobian
+        Eigen::MatrixXd Jf_x = fx.get_jacobian(theta_dbl);
+        hybrj_functor_solver<FS, F, double, double>
+          fy(fs, f, theta_dbl, value_of(y), dat, dat_int, msgs, false);
+        Eigen::MatrixXd Jf_y = fy.get_jacobian(value_of(y));
+
+        Jx_y_ = - stan::math::mdivide_left(Jf_x, Jf_y);
+      }
+
+      void chain() {
+        for (int i = 0; i < x_size_; i++)
+          for (int j = 0; j < y_size_; j++)
```

Comment:
oh, wow. I've always been curious about how and why column vs row major makes a difference. I guess it comes down to how memory is allocated, since the memory itself is not stored in a matrix form. Anyways, this does the trick, right?
```
for (int j = 0; j < y_size_; j++)
  for (int i = 0; i < x_size_; i++)
     y_[j]->adj_ += theta_[i]->adj_ * Jx_y_(i, j);
```